src/storage/v2/replication/replication_storage_state.cpp

```diff
@@ -14,72 +14,95 @@
+    -> std::vector<std::optional<ReplicaStream>> {
+  std::vector<std::optional<ReplicaStream>> replica_streams;
+  replication_clients_.WithLock([&, db_acc = std::move(db_acc)](auto &clients) mutable {
     for (auto &client : clients) {
-      client->StartTransactionReplication(seq_num, storage, std::move(db_acc));
+      replica_streams.emplace_back(client->StartTransactionReplication(seq_num, storage, std::move(db_acc)));
     }
   });
+  return replica_streams;
 }
 
-void ReplicationStorageState::AppendDelta(const Delta &delta, const Vertex &vertex, uint64_t timestamp) {
+void ReplicationStorageState::AppendDelta(const Delta &delta, const Vertex &vertex, uint64_t timestamp,
+                                          std::span<std::optional<ReplicaStream>> replica_streams) {
   replication_clients_.WithLock([&](auto &clients) {
-    for (auto &client : clients) {
-      client->IfStreamingTransaction([&](auto &stream) { stream.AppendDelta(delta, vertex, timestamp); });
+    for (int i = 0; i < clients.size(); i++) {
+      auto &client = clients[i];
+      auto &replica_stream = replica_streams[i];
+      client->IfStreamingTransaction([&](auto &stream) { stream.AppendDelta(delta, vertex, timestamp); },
+                                     replica_stream);
     }
   });
 }
 
-void ReplicationStorageState::AppendDelta(const Delta &delta, const Edge &edge, uint64_t timestamp) {
+void ReplicationStorageState::AppendDelta(const Delta &delta, const Edge &edge, uint64_t timestamp,
+                                          std::span<std::optional<ReplicaStream>> replica_streams) {
   replication_clients_.WithLock([&](auto &clients) {
-    for (auto &client : clients) {
-      client->IfStreamingTransaction([&](auto &stream) { stream.AppendDelta(delta, edge, timestamp); });
+    for (int i = 0; i < clients.size(); i++) {
+      auto &client = clients[i];
+      auto &replica_stream = replica_streams[i];
+      client->IfStreamingTransaction([&](auto &stream) { stream.AppendDelta(delta, edge, timestamp); }, replica_stream);
     }
   });
 }
 void ReplicationStorageState::AppendOperation(durability::StorageMetadataOperation operation, LabelId label,
                                               const std::set<PropertyId> &properties, const LabelIndexStats &stats,
                                               const LabelPropertyIndexStats &property_stats,
-                                              uint64_t final_commit_timestamp) {
+                                              uint64_t final_commit_timestamp,
+                                              std::span<std::optional<ReplicaStream>> replica_streams) {
   replication_clients_.WithLock([&](auto &clients) {
-    for (auto &client : clients) {
-      client->IfStreamingTransaction([&](auto &stream) {
-        stream.AppendOperation(operation, label, properties, stats, property_stats, final_commit_timestamp);
-      });
+    for (int i = 0; i < clients.size(); i++) {
```

Comment:
ranges-v3 enumerate or zip

Reply:
Also with raw loops, be intentional: `for (int i = 0; i != clients.size(); ++i)`